### ISSUE TITLE
Why not running fluent as root?

### ISSUE THREAD

**Issue by alexanderkiel** [fluent/fluentd-docker-image#48]
I have the problem that fluent can't write the pos files of the tail plugin into the volume I mount for tailing. The reason is that fluentd doesn't run as root inside the container. Is there a reason for not running fluentd as root?

**Comment by jescarri** [comment]
I guess is because the gems, I do not see why it can't be run as root, I have the same problem.

**Comment by repeatedly** [comment]
It follows Docker best practice: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#/user

How other product handle volume permission issue?
All product runs as a root?

**Comment by jescarri** [comment]
If you are trying to use fluentd-plugin-systemd it requires access to the journal file, and that one is owned by r00t.

**Comment by repeatedly** [comment]
Hmm... how about setting "USER root" in your dockerfile?

**Comment by jescarri** [comment]
Yep, it will run the next commands or entry point as root, which is the same as not using the fluent user.

**Comment by alexanderkiel** [comment]
I'm not aware of any containers which run the process as non-root. I've just build my own fluent image.

**Comment by ferminromero** [comment]
@repeatedly -- Setting `USER root` does not work because the entrypoint actually intentionally switches the user to fluentd.

https://github.com/fluent/fluentd-docker-image/blob/master/v0.14/alpine/entrypoint.sh#L18

This is unexpected and annoying.

**Comment by repeatedly** [comment]
Why need `USER root`?

**Comment by ferminromero** [comment]
We are running this in Kubernetes and mounting all of our container's logs into the fluentd container.
These are mounted as root and it isn't possible to change the permissions they are mounted with.

I understand that this container expects files to be mounted to `/fluentd/log/` but then we have to customize our kubernetes_metadata_filter plugin... which is annoying.

It also doesn't look like we can add the fluentd user to the root group in our Dockerfile because it doesn't exist... And then the entrypoint deletes and recreates the user.

I'm sure there _is_ a workaround, but at this point, we're just going to fork and build our own container.

Long story short... This doesn't work well with kubernetes right now IMO.

**Comment by repeatedly** [comment]
I see. I'm not familiar with k8s operation...

@ValFadeev Do you have any insight?

**Comment by edsiper** [comment]
For specific Kubernetes use case we have official images at Docker hub:

https://hub.docker.com/r/fluent/fluentd-kubernetes-daemonset/

there are debian and alpine images integrated for s3, loggly, cloudwatch, etc.

**Comment by ferminromero** [comment]
@edsiper -- We're logging to rsyslog, so those images won't work for us (I can see if we can make an upstream contribution at a later date). 

We actually used those images as a base for ours, and we can't figure out why your images are able to access `/var/log` -- The only thing that jumped out was `USER root` in the Dockerfile, which doesn't seem to work for us (and doesn't seem like it _should_ work given the `su-exec fluentd` in the entrypoint)

------

Alright, our workaround is as follows...

In our image that uses `FROM fluent/fluentd:v0.14.17`, we copy in a custom `entrypoint.sh` script that changes the last line to `exec "$@"`

**Comment by edsiper** [comment]
@ferminromero if you are deploying on Kubernetes you should use our daemonset debian images, you can use our templates to add the rsyslog configuration on it. 

The images can consume logs without problems.

**Comment by ferminromero** [comment]
In other words, we should extend the `fluentd-kubernetes-daemonset` repository to support rsyslog? (Just want to make sure that isn't already a supported configuration)

**Comment by edsiper** [comment]
Yes, you should extend that one, because it's already set with filters and proper plugins (use Debian based images)

**Comment by ferminromero** [comment]
@edsiper -- We sent out a PR -- https://github.com/fluent/fluentd-kubernetes-daemonset/pull/32
Any reason the debian images are preferred (just curious)

**Comment by edsiper** [comment]
@ferminromero thanks, I will review the PR shortly. 

Now between Alpine and Debian, we ran into many issues because of Musl C library, that's the one used by Alpine and some things are not fully compatible when using specific libc calls.

In the other hand having Jemalloc memory allocator is highly desired to reduce memory fragmentation and Alpine + Jemalloc are not working properly together.

**Comment by j-davies** [comment]
@ferminromero Did you still need your workaround:

> In our image that uses FROM fluent/fluentd:v0.14.17, we copy in a custom entrypoint.sh script that changes the last line to exec "$@"

When running the daemonset you defined in [your pr](https://github.com/fluent/fluentd-kubernetes-daemonset/pull/32)?

I am attempting to add the bigquery plugin to the fluentd-kubernetes-daemonset in the similar way you tried to add rsyslog, but am having the same issue as you with flunetd running as user fluent, and not root on k8, and not having access to log directories.

**Comment by xinghusc** [comment]
We've encountered the same situation in our project and we do want to use root instead of well confined user "fluent". The workaround we did was to [override the entrypoint](https://stackoverflow.com/questions/41207522/docker-override-or-remove-entrypoint-from-a-base-image) of the base fluentd image, so we don't need to fork our own version of the base image.

**Comment by klo2k** [comment]
> We've encountered the same situation in our project and we do want to use root instead of well confined user "fluent". The workaround we did was to [override the entrypoint](https://stackoverflow.com/questions/41207522/docker-override-or-remove-entrypoint-from-a-base-image) of the base fluentd image, so we don't need to fork our own version of the base image.

Looks like we can mount our custom entry-point with `--volume` + override with `--entrypoint`.

e.g.:
```
# Custom entry-point script
$ cat /tmp/test-entrypoint.sh 
#!/bin/sh
echo "Running as $(whoami)"
"$@"

# Run with custom entry-point script
$ docker run --entrypoint=/tmp/test-entrypoint.sh --rm -it --volume /tmp/test-entrypoint.sh:/tmp/test-entrypoint.sh fluent/fluentd sh
Running as root
/ # 
```

**Comment by klo2k** [comment]
...and with the latest image "edge" (sorry I'm new to fluentd...), use `--user root`

```
$ docker run --user root --rm -it fluent/fluentd:edge whoami
root
```

**Comment by ben-bourdin451** [comment]
This is an old thread, however I want some thoughts on this issue as I recently ran into this problem when mounting a volume using the base image.

Not having to run fluentd as root is best practice however IMO, 1 of 2 improvements could be made:

1. Adding user & group as the first instruction to ensure consistent uid & gid
This is being done in [Redis for example](https://github.com/docker-library/redis/blob/master/Dockerfile.template#L4)

2. Or perhaps better yet, explicitly assign a known uid & gid to the fluent user & group so this can be re-used for file & mount permissions

Currently we are using the workaround that @klo2k mentioned right after mounting a volume:
```sh
docker run --rm \
	-u root \
	-v ${MOUNT}:${TARGET} \
	fluent/fluentd:edge \
	chown -R fluent:fluent ${TARGET}
```
